### ISSUE TITLE
Cluster version skew rule violation

### ISSUE THREAD

**Issue by balous** [kubermatic/kubermatic#5949]
**User Story**
I've tried to upgrade user cluster by editing its resource and changing `version: 1.17.9` -> `version: 1.18.6`. I've observed that the control plane pods began to update instantly - apiserver, controller manager and scheduler in parallel.

This approach violates the [Kubernetes version and version skew support policy](https://kubernetes.io/docs/setup/release/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager) which states that apiserver must not be newer that the other control plane components.

Parallel upgrade leads to a situation that (for example) upgraded CM pod communicates with apiserver pod that is still waiting for upgrade. This can lead to anything between strange errors in logs and momentary cluster misfunction.

**Acceptance criteria**

The solution is to upgrade apiserver first, wait for completion and than proceeding to the other components.

**Comment by kron4eg** [comment]
>which states that apiserver must not be newer that the other control plane components.

I'm confused. Quote from your link:
>`kube-controller-manager`, `kube-scheduler`, and `cloud-controller-manager` must not be newer than the `kube-apiserver`

The exact opposite.

The implementation of the upgrade is done so that all components are reconciled (thus restarted) at once, there are no "newer" / "older" components.

IMHO, this issue is invalid.

**Comment by balous** [comment]
This is because api server and the other control plane components run in multiple instances and rolling update is configured for them. The update eventually comes to situation where a mixture of old and new versions is running - for all of the control plane components.

For example when we update 1.17 -> 1.18 in such a moment following component mix is running:

- apiserver 1.17
- apiserver 1.18
- controller manager 1.17
- controller manager 1.18

It is possible that, upon its startup, the controller manager 1.18 connects to apiserver 1.17 which is still waiting for update. Thus the version skew policy is violated.

Kubernetes are designed so that the cluster is functional even during the update. In order to achieve this the developers defined these rules, designed apiserver to be backwards compatible and they (hopefully :) test such scenarios.

The update must be done i a way that apiserver update is completed before the process advances to updating the rest of the cluster.

**Comment by scheeles** [comment]
Agree with @balous currently we don't have a phase upgrade. 

We had a `MasterUpdatePhase` in the past https://github.com/kubermatic/kubermatic/blob/26ac0519f384ced6eb0d40c52827f296b103db5f/api/pkg/crd/kubermatic/v1/cluster.go#L44

somehow it was removed in this https://github.com/kubermatic/kubermatic/pull/1095 

We need to reimplement it again.

**Comment by xrstf** [comment]
Meetings notes:

So our current plan is to keep this pretty simple: The upgrade controller will check the status of the Deployments for the control-plane components (apiserver, cm, scheduler) and be more delicate when updating them.

So when a user clicks on upgrade, the controller will

* check the existing Deployments and their status
  * are they Ready?
  * what version are they running (we can put this in a label/annotation so we do not have to parse the Docker image tag)
* if all Deployments are ready, upgrade the apiserver Deployment
* wait for it to completely roll out (check this based on the apiserver Deployment status)
* once the apiserver is rolled out, upgrade cm/scheduler Deployments
* rinse and repeat

We do not intend to build a generic solution for deferred/dependent resources.
